spec/services/bulk_upload/lettings/year2023/row_parser_spec.rb

```diff
@@ -806,8 +818,50 @@
       context "when over 12 characters" do
         let(:attributes) { { bulk_upload:, field_18: "1234567890123" } }
 
-        it "has errors on the field" do
-          expect(parser.errors[:field_18]).to be_present
+        it "adds an appropriate error" do
+          expect(parser.errors[:field_18]).to eql(["UPRN must be 12 digits or less"])
+        end
+      end
+
+      context "when neither UPRN nor address fields are given" do
+        let(:attributes) do
+          {
+            bulk_upload:,
+            field_1: "1",
+          }
+        end
+
+        it "adds appropriate errors" do
+          expect(parser.errors[:field_18]).to eql(["You must answer UPRN"])
```

Comment:
Is validations.not_answered also being triggered? (in the row parser's validate_nulls method)

Reply:
Thanks for looking at this @natdeanlewissoftwire! For the record, no it wasn't being triggered (that was my first thought too though). I've had a look at why and this is what I've worked out:

It wasn't being triggered because when the UPRN is blank in a bulk upload, there's a line in the row parser which automatically sets `uprn_known` as 0. Then the answer to the UPRN question is inferred (as "Not known") as can be seen in `app/models/form/lettings/questions/uprn.rb`, and because it's inferred it's understood to be completed. So the final if statement in the `validate_nulls` method doesn't get reached because it ignores completed questions. This is fine because the UPRN question is allowed to be null - that's why the UPRN known? question exists. The reason we want to see a "You must answer UPRN" error message is because UPRN can only be null if certain other address fields (address line 1 and town/city) are not null.